Add missing `UserBgtaskEvent` implementation

## Summary  

```
2025-05-15 07:38:26.097 DEBUG ai.backend.manager.plugin.error_monitor [501670] collected an error log [ERROR] "AttributeError: 'UUID' object has no attribute 'replace'" from manager
2025-05-15 07:38:26.098 ERROR ai.backend.manager.server [501670] Uncaught exception in HTTP request handlers AttributeError("'UUID' object has no attribute 'replace'")
Traceback (most recent call last):
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/server.py", line 304, in exception_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/server.py", line 291, in api_middleware
    resp = await _handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/common/metrics/http.py", line 39, in metric_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/api/auth.py", line 593, in auth_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/aiotools/func.py", line 23, in wrapped
    return await coro(*args, *cargs, **kwargs, **ckwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/api/ratelimit.py", line 73, in rlim_middleware
    response = await handler(request)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/dist/export/python/virtualenvs/python-default/3.13.3/lib/python3.13/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/api/manager.py", line 80, in wrapped
    return await handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/api/auth.py", line 600, in wrapped
    return await handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/api/utils.py", line 211, in wrapped
    return await handler(request, checked_params, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jopemachine/.local/backend.ai/repos/feat_add_auditlog_api/src/ai/backend/manager/api/events.py", line 183, in push_background_task_events
    task_id = uuid.UUID(params["task_id"])
  File "/home/jopemachine/.pyenv/versions/3.13.3/lib/python3.13/uuid.py", line 178, in __init__
    hex = hex.replace('urn:', '').replace('uuid:', '')
          ^^^^^^^^^^^
AttributeError: 'UUID' object has no attribute 'replace'
```

--- a/changes/4404.fix.md
+++ b/changes/4404.fix.md
@@ -1,0 +1,1 @@
+Add missing `UserBgtaskEvent` implementation.

--- a/src/ai/backend/common/bgtask/bgtask.py
+++ b/src/ai/backend/common/bgtask/bgtask.py
@@ -213,6 +213,7 @@ class BackgroundTaskManager:
             # The task ID is invalid or represents a task completed more than timeout.
             raise BgtaskNotFoundError("No such background task.")
 
+        task_info_dict["status"] = BgtaskStatus(task_info_dict["status"])
         task_info = _BgTaskInfo(**task_info_dict)
         if not task_info.status.finished():
             return None

--- a/src/ai/backend/common/bgtask/bgtask.py
+++ b/src/ai/backend/common/bgtask/bgtask.py
@@ -14,6 +14,7 @@ from typing import (
     Mapping,
     Optional,
     Protocol,
+    Self,
     TypeAlias,
     Union,
 )
@@ -28,7 +29,6 @@ from ai.backend.logging import BraceStyleAdapter
 from .. import redis_helper
 from ..events.bgtask import (
     BaseBgtaskEvent,
-    BgtaskAlreadyDoneEvent,
     BgtaskCancelledEvent,
     BgtaskDoneEvent,
     BgtaskFailedEvent,
@@ -56,7 +56,7 @@ _MAX_BGTASK_ARCHIVE_PERIOD: Final = 86400  # 24  hours
 
 
 @dataclass
-class _BgTaskInfo:
+class BgTaskInfo:
     status: BgtaskStatus
     msg: str
     started_at: str
@@ -65,7 +65,7 @@ class _BgTaskInfo:
     total: str = "0"
 
     @classmethod
-    def started(cls, msg: str = "") -> _BgTaskInfo:
+    def started(cls, msg: str = "") -> Self:
         now = str(time.time())
         return cls(
             status=BgtaskStatus.STARTED,
@@ -77,7 +77,7 @@ class _BgTaskInfo:
         )
 
     @classmethod
-    def finished(cls, status: BgtaskStatus, msg: str = "") -> _BgTaskInfo:
+    def finished(cls, status: BgtaskStatus, msg: str = "") -> Self:
         now = str(time.time())
         return cls(
             status=status,
@@ -199,10 +199,10 @@ class BackgroundTaskManager:
         self._metric_observer = bgtask_observer
         self._dict_lock = asyncio.Lock()
 
-    async def fetch_last_finished_event(
+    async def fetch_bgtask_info(
         self,
         task_id: uuid.UUID,
-    ) -> Optional[BaseBgtaskEvent]:
+    ) -> BgTaskInfo:
         tracker_key = _tracker_id(task_id)
         task_info_dict = await redis_helper.execute(
             self._redis_client,
@@ -214,17 +214,8 @@ class BackgroundTaskManager:
             raise BgtaskNotFoundError("No such background task.")
 
         task_info_dict["status"] = BgtaskStatus(task_info_dict["status"])
-        task_info = _BgTaskInfo(**task_info_dict)
-        if not task_info.status.finished():
-            return None
-
-        return BgtaskAlreadyDoneEvent(
-            task_id=task_id,
-            message=task_info.msg,
-            task_status=task_info.status,
-            current=task_info.current,
-            total=task_info.total,
-        )
+        task_info = BgTaskInfo(**task_info_dict)
+        return task_info
 
     async def start(
         self,
@@ -259,11 +250,11 @@ class BackgroundTaskManager:
 
         async def _pipe_builder(r: Redis) -> Pipeline:
             pipe = r.pipeline()
-            task_info: _BgTaskInfo
+            task_info: BgTaskInfo
             if status.finished():
-                task_info = _BgTaskInfo.finished(status=status, msg=msg)
+                task_info = BgTaskInfo.finished(status=status, msg=msg)
             else:
-                task_info = _BgTaskInfo.started(msg=msg)
+                task_info = BgTaskInfo.started(msg=msg)
             mapping = task_info.to_dict()
             pipe.hset(tracker_key, mapping=mapping)
             pipe.expire(tracker_key, _MAX_BGTASK_ARCHIVE_PERIOD)

--- a/src/ai/backend/common/events/bgtask.py
+++ b/src/ai/backend/common/events/bgtask.py
@@ -182,8 +182,6 @@ class BgtaskCancelledEvent(BaseBgtaskDoneEvent):
         return UserBgtaskCancelledEvent(
             task_id=str(self.task_id),
             message=str(self.message),
-            current_progress=0,
-            total_progress=0,
         )
 
 
@@ -203,8 +201,6 @@ class BgtaskFailedEvent(BaseBgtaskDoneEvent):
         return UserBgtaskFailedEvent(
             task_id=str(self.task_id),
             message=str(self.message),
-            current_progress=0,
-            total_progress=0,
         )
 
 
@@ -244,6 +240,5 @@ class BgtaskPartialSuccessEvent(BaseBgtaskDoneEvent):
         return UserBgtaskPartialSuccessEvent(
             task_id=str(self.task_id),
             message=str(self.message),
-            current_progress=0,
-            total_progress=0,
+            errors=self.errors,
         )

--- a/src/ai/backend/common/events/bgtask.py
+++ b/src/ai/backend/common/events/bgtask.py
@@ -5,6 +5,14 @@ from typing import Optional, override
 
 from ai.backend.common.bgtask.types import BgtaskStatus
 from ai.backend.common.events.dispatcher import AbstractEvent, EventDomain
+from ai.backend.common.events.user_event.user_bgtask_event import (
+    UserBgtaskAlreadyDoneEvent,
+    UserBgtaskCancelledEvent,
+    UserBgtaskDoneEvent,
+    UserBgtaskFailedEvent,
+    UserBgtaskPartialSuccessEvent,
+    UserBgtaskUpdatedEvent,
+)
 from ai.backend.common.events.user_event.user_event import UserEvent
 from ai.backend.common.exception import UnreachableError
 
@@ -21,10 +29,6 @@ class BaseBgtaskEvent(AbstractEvent, ABC):
     @override
     def domain_id(self) -> Optional[str]:
         return str(self.task_id)
-
-    @override
-    def user_event(self) -> Optional[UserEvent]:
-        return None
 
     @abstractmethod
     def status(self) -> BgtaskStatus:
@@ -62,6 +66,15 @@ class BgtaskUpdatedEvent(BaseBgtaskEvent):
     @override
     def status(self) -> BgtaskStatus:
         return BgtaskStatus.STARTED
+
+    @override
+    def user_event(self) -> Optional[UserEvent]:
+        return UserBgtaskUpdatedEvent(
+            task_id=str(self.task_id),
+            message=str(self.message),
+            current_progress=self.current_progress,
+            total_progress=self.total_progress,
+        )
 
 
 @dataclass
@@ -103,6 +116,15 @@ class BgtaskDoneEvent(BaseBgtaskDoneEvent):
     def status(self) -> BgtaskStatus:
         return BgtaskStatus.DONE
 
+    @override
+    def user_event(self) -> Optional[UserEvent]:
+        return UserBgtaskDoneEvent(
+            task_id=str(self.task_id),
+            message=str(self.message),
+            current_progress=0,
+            total_progress=0,
+        )
+
 
 @dataclass
 class BgtaskAlreadyDoneEvent(BaseBgtaskEvent):
@@ -134,6 +156,15 @@ class BgtaskAlreadyDoneEvent(BaseBgtaskEvent):
     def status(self) -> BgtaskStatus:
         return self.task_status
 
+    @override
+    def user_event(self) -> Optional[UserEvent]:
+        return UserBgtaskAlreadyDoneEvent(
+            task_id=str(self.task_id),
+            message=str(self.message),
+            current_progress=float(self.current),
+            total_progress=float(self.total),
+        )
+
 
 @dataclass
 class BgtaskCancelledEvent(BaseBgtaskDoneEvent):
@@ -146,6 +177,15 @@ class BgtaskCancelledEvent(BaseBgtaskDoneEvent):
     def status(self) -> BgtaskStatus:
         return BgtaskStatus.CANCELLED
 
+    @override
+    def user_event(self) -> Optional[UserEvent]:
+        return UserBgtaskCancelledEvent(
+            task_id=str(self.task_id),
+            message=str(self.message),
+            current_progress=0,
+            total_progress=0,
+        )
+
 
 @dataclass
 class BgtaskFailedEvent(BaseBgtaskDoneEvent):
@@ -157,6 +197,15 @@ class BgtaskFailedEvent(BaseBgtaskDoneEvent):
     @override
     def status(self) -> BgtaskStatus:
         return BgtaskStatus.FAILED
+
+    @override
+    def user_event(self) -> Optional[UserEvent]:
+        return UserBgtaskFailedEvent(
+            task_id=str(self.task_id),
+            message=str(self.message),
+            current_progress=0,
+            total_progress=0,
+        )
 
 
 @dataclass
@@ -189,3 +238,12 @@ class BgtaskPartialSuccessEvent(BaseBgtaskDoneEvent):
     def status(self) -> BgtaskStatus:
         # TODO: When client side is ready, we can change this to `TaskStatus.PARTIAL_SUCCESS`
         return BgtaskStatus.DONE
+
+    @override
+    def user_event(self) -> Optional[UserEvent]:
+        return UserBgtaskPartialSuccessEvent(
+            task_id=str(self.task_id),
+            message=str(self.message),
+            current_progress=0,
+            total_progress=0,
+        )

--- a/src/ai/backend/common/events/bgtask.py
+++ b/src/ai/backend/common/events/bgtask.py
@@ -121,8 +121,6 @@ class BgtaskDoneEvent(BaseBgtaskDoneEvent):
         return UserBgtaskDoneEvent(
             task_id=str(self.task_id),
             message=str(self.message),
-            current_progress=0,
-            total_progress=0,
         )
 
 

--- a/src/ai/backend/common/events/bgtask.py
+++ b/src/ai/backend/common/events/bgtask.py
@@ -10,7 +10,6 @@ from ai.backend.common.events.user_event.user_bgtask_event import (
     UserBgtaskCancelledEvent,
     UserBgtaskDoneEvent,
     UserBgtaskFailedEvent,
-    UserBgtaskPartialSuccessEvent,
     UserBgtaskUpdatedEvent,
 )
 from ai.backend.common.events.user_event.user_event import UserEvent
@@ -235,8 +234,8 @@ class BgtaskPartialSuccessEvent(BaseBgtaskDoneEvent):
 
     @override
     def user_event(self) -> Optional[UserEvent]:
-        return UserBgtaskPartialSuccessEvent(
+        # TODO: When client side is ready, we can change this to `UserBgtaskPartialSuccessEvent`
+        return UserBgtaskDoneEvent(
             task_id=str(self.task_id),
             message=str(self.message),
-            errors=self.errors,
         )

--- a/src/ai/backend/common/events/hub/propagators/bgtask.py
+++ b/src/ai/backend/common/events/hub/propagators/bgtask.py
@@ -3,9 +3,10 @@ import logging
 import uuid
 from typing import AsyncIterator, Optional, Protocol
 
+from ai.backend.common.bgtask.bgtask import BgTaskInfo
 from ai.backend.logging.utils import BraceStyleAdapter
 
-from ...bgtask import BaseBgtaskEvent
+from ...bgtask import BgtaskAlreadyDoneEvent
 from ...dispatcher import AbstractEvent
 from ..hub import EventPropagator
 
@@ -13,13 +14,13 @@ log = BraceStyleAdapter(logging.getLogger(__spec__.name))
 
 
 class BgtaskLastDoneEventFetcher(Protocol):
-    async def fetch_last_finished_event(
+    async def fetch_bgtask_info(
         self,
         task_id: uuid.UUID,
-    ) -> Optional[BaseBgtaskEvent]:
+    ) -> BgTaskInfo:
         """
-        Fetch the last finished event for a given task ID.
-        This method should be implemented by the class that uses this protocol.
+        Fetch the background task information for a given task ID.
+        This method is used to retrieve the last status of a background task.
         """
         ...
 
@@ -54,13 +55,19 @@ class BgtaskPropagator(EventPropagator):
         If the last event is not None, it yields that event.
         Then, it enters a loop to receive events from the queue until closed.
         """
-        last_event = await self._last_done_event_fetcher.fetch_last_finished_event(task_id)
-        if last_event is not None:
+        bgtask_info = await self._last_done_event_fetcher.fetch_bgtask_info(task_id)
+        if bgtask_info.status.finished():
             log.debug(
-                "Yielding last finished event: {}",
-                last_event,
+                "Yielding already finished event: {}",
+                bgtask_info,
             )
-            yield last_event
+            yield BgtaskAlreadyDoneEvent(
+                task_id=task_id,
+                message=bgtask_info.msg,
+                task_status=bgtask_info.status,
+                current=bgtask_info.current,
+                total=bgtask_info.total,
+            )
             return
         while not self._closed:
             event = await self._queue.get()

--- a/src/ai/backend/common/events/user_event/user_bgtask_event.py
+++ b/src/ai/backend/common/events/user_event/user_bgtask_event.py
@@ -23,3 +23,123 @@ class UserBgTaskEvent(UserEvent):  # TODO: Need to separate UserPartialSuccessEv
     @override
     def user_event_mapping(self) -> Mapping[str, Any]:
         return asdict(self)
+
+
+@dataclass
+class UserBgtaskUpdatedEvent(UserEvent):
+    task_id: str
+    message: str
+    current_progress: float
+    total_progress: float
+
+    @override
+    def event_name(self) -> Optional[str]:
+        return "bgtask_updated"
+
+    @override
+    def retry_count(self) -> Optional[int]:
+        return 5
+
+    @override
+    def user_event_mapping(self) -> Mapping[str, Any]:
+        return asdict(self)
+
+
+@dataclass
+class UserBgtaskDoneEvent(UserEvent):
+    task_id: str
+    message: str
+    current_progress: float
+    total_progress: float
+
+    @override
+    def event_name(self) -> Optional[str]:
+        return "bgtask_done"
+
+    @override
+    def retry_count(self) -> Optional[int]:
+        return None
+
+    @override
+    def user_event_mapping(self) -> Mapping[str, Any]:
+        return asdict(self)
+
+
+@dataclass
+class UserBgtaskPartialSuccessEvent(UserEvent):
+    task_id: str
+    message: str
+    current_progress: float
+    total_progress: float
+
+    @override
+    def event_name(self) -> Optional[str]:
+        return "bgtask_partial_success"
+
+    @override
+    def retry_count(self) -> Optional[int]:
+        return None
+
+    @override
+    def user_event_mapping(self) -> Mapping[str, Any]:
+        return asdict(self)
+
+
+@dataclass
+class UserBgtaskCancelledEvent(UserEvent):
+    task_id: str
+    message: str
+    current_progress: float
+    total_progress: float
+
+    @override
+    def event_name(self) -> Optional[str]:
+        return "bgtask_cancelled"
+
+    @override
+    def retry_count(self) -> Optional[int]:
+        return None
+
+    @override
+    def user_event_mapping(self) -> Mapping[str, Any]:
+        return asdict(self)
+
+
+@dataclass
+class UserBgtaskFailedEvent(UserEvent):
+    task_id: str
+    message: str
+    current_progress: float
+    total_progress: float
+
+    @override
+    def event_name(self) -> Optional[str]:
+        return "bgtask_failed"
+
+    @override
+    def retry_count(self) -> Optional[int]:
+        return None
+
+    @override
+    def user_event_mapping(self) -> Mapping[str, Any]:
+        return asdict(self)
+
+
+@dataclass
+class UserBgtaskAlreadyDoneEvent(UserEvent):
+    task_id: str
+    message: str
+    current_progress: float
+    total_progress: float
+
+    @override
+    def event_name(self) -> Optional[str]:
+        return "bgtask_already_done"
+
+    @override
+    def retry_count(self) -> Optional[int]:
+        return None
+
+    @override
+    def user_event_mapping(self) -> Mapping[str, Any]:
+        return asdict(self)

--- a/src/ai/backend/common/events/user_event/user_bgtask_event.py
+++ b/src/ai/backend/common/events/user_event/user_bgtask_event.py
@@ -69,8 +69,7 @@ class UserBgtaskDoneEvent(UserEvent):
 class UserBgtaskPartialSuccessEvent(UserEvent):
     task_id: str
     message: str
-    current_progress: float
-    total_progress: float
+    errors: list[str]
 
     @override
     def event_name(self) -> Optional[str]:
@@ -89,8 +88,6 @@ class UserBgtaskPartialSuccessEvent(UserEvent):
 class UserBgtaskCancelledEvent(UserEvent):
     task_id: str
     message: str
-    current_progress: float
-    total_progress: float
 
     @override
     def event_name(self) -> Optional[str]:
@@ -109,8 +106,6 @@ class UserBgtaskCancelledEvent(UserEvent):
 class UserBgtaskFailedEvent(UserEvent):
     task_id: str
     message: str
-    current_progress: float
-    total_progress: float
 
     @override
     def event_name(self) -> Optional[str]:

--- a/src/ai/backend/common/events/user_event/user_bgtask_event.py
+++ b/src/ai/backend/common/events/user_event/user_bgtask_event.py
@@ -71,7 +71,7 @@ class UserBgtaskPartialSuccessEvent(UserEvent):
 
     @override
     def event_name(self) -> Optional[str]:
-        return "bgtask_partial_success"
+        return "bgtask_done"
 
     @override
     def retry_count(self) -> Optional[int]:

--- a/src/ai/backend/common/events/user_event/user_bgtask_event.py
+++ b/src/ai/backend/common/events/user_event/user_bgtask_event.py
@@ -49,8 +49,6 @@ class UserBgtaskUpdatedEvent(UserEvent):
 class UserBgtaskDoneEvent(UserEvent):
     task_id: str
     message: str
-    current_progress: float
-    total_progress: float
 
     @override
     def event_name(self) -> Optional[str]:

--- a/src/ai/backend/manager/api/events.py
+++ b/src/ai/backend/manager/api/events.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 import asyncio
 import logging
+import uuid
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -179,7 +180,7 @@ async def push_background_task_events(
     params: Mapping[str, Any],
 ) -> web.StreamResponse:
     root_ctx: RootContext = request.app["_root.context"]
-    task_id = params["task_id"]
+    task_id: uuid.UUID = params["task_id"]
     access_key = request["keypair"]["access_key"]
     log.info("PUSH_BACKGROUND_TASK_EVENTS (ak:{}, t:{})", access_key, task_id)
     async with sse_response(request) as resp:
@@ -201,6 +202,7 @@ async def push_background_task_events(
                     event=user_event.event_name(),
                     retry=user_event.retry_count(),
                 )
+                await propagator.close()
             await resp.send(dump_json_str({}), event="server_close")
         finally:
             root_ctx.event_hub.unregister_event_propagator(propagator.id())

--- a/src/ai/backend/manager/api/events.py
+++ b/src/ai/backend/manager/api/events.py
@@ -202,7 +202,8 @@ async def push_background_task_events(
                     event=user_event.event_name(),
                     retry=user_event.retry_count(),
                 )
-                await propagator.close()
+                # TODO: How to properly terminate a propagator?
+                # await propagator.close()
             await resp.send(dump_json_str({}), event="server_close")
         finally:
             root_ctx.event_hub.unregister_event_propagator(propagator.id())

--- a/src/ai/backend/manager/api/events.py
+++ b/src/ai/backend/manager/api/events.py
@@ -202,8 +202,6 @@ async def push_background_task_events(
                     event=user_event.event_name(),
                     retry=user_event.retry_count(),
                 )
-                # TODO: How to properly terminate a propagator?
-                # await propagator.close()
             await resp.send(dump_json_str({}), event="server_close")
         finally:
             root_ctx.event_hub.unregister_event_propagator(propagator.id())

--- a/src/ai/backend/manager/api/events.py
+++ b/src/ai/backend/manager/api/events.py
@@ -2,7 +2,6 @@ from __future__ import annotations
 
 import asyncio
 import logging
-import uuid
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -180,7 +179,7 @@ async def push_background_task_events(
     params: Mapping[str, Any],
 ) -> web.StreamResponse:
     root_ctx: RootContext = request.app["_root.context"]
-    task_id = uuid.UUID(params["task_id"])
+    task_id = params["task_id"]
     access_key = request["keypair"]["access_key"]
     log.info("PUSH_BACKGROUND_TASK_EVENTS (ak:{}, t:{})", access_key, task_id)
     async with sse_response(request) as resp:

--- a/src/ai/backend/manager/event_dispatcher/dispatch.py
+++ b/src/ai/backend/manager/event_dispatcher/dispatch.py
@@ -22,14 +22,14 @@ def _dispatch_bgtask_events(
     Register event dispatchers for background task events.
     """
     event_dispatcher.subscribe(BgtaskUpdatedEvent, None, dispatcher.propagate_event)
-    event_dispatcher.subscribe(BgtaskDoneEvent, None, dispatcher.propagate_event)
+    event_dispatcher.subscribe(BgtaskDoneEvent, None, dispatcher.propagate_event_with_close)
     event_dispatcher.subscribe(
         BgtaskPartialSuccessEvent,
         None,
-        dispatcher.propagate_event,
+        dispatcher.propagate_event_with_close,
     )
-    event_dispatcher.subscribe(BgtaskCancelledEvent, None, dispatcher.propagate_event)
-    event_dispatcher.subscribe(BgtaskFailedEvent, None, dispatcher.propagate_event)
+    event_dispatcher.subscribe(BgtaskCancelledEvent, None, dispatcher.propagate_event_with_close)
+    event_dispatcher.subscribe(BgtaskFailedEvent, None, dispatcher.propagate_event_with_close)
 
 
 @dataclass

--- a/src/ai/backend/manager/event_dispatcher/propagator.py
+++ b/src/ai/backend/manager/event_dispatcher/propagator.py
@@ -1,6 +1,7 @@
 from ai.backend.common.events.dispatcher import AbstractEvent
 from ai.backend.common.events.hub.hub import EventHub
 from ai.backend.common.types import AgentId
+from ai.backend.manager.errors.exceptions import InternalServerError
 
 
 class PropagatorEventDispatcher:
@@ -26,5 +27,5 @@ class PropagatorEventDispatcher:
         await self._event_hub.propagate_event(event)
         domain_id = event.domain_id()
         if domain_id is None:
-            raise AssertionError("Event domain ID is None")
+            raise InternalServerError("Event domain ID is None")
         await self._event_hub.close_by_alias(event.event_domain(), domain_id)

--- a/src/ai/backend/manager/event_dispatcher/propagator.py
+++ b/src/ai/backend/manager/event_dispatcher/propagator.py
@@ -16,3 +16,15 @@ class PropagatorEventDispatcher:
         event: AbstractEvent,
     ) -> None:
         await self._event_hub.propagate_event(event)
+
+    async def propagate_event_with_close(
+        self,
+        context: None,
+        source: AgentId,
+        event: AbstractEvent,
+    ) -> None:
+        await self._event_hub.propagate_event(event)
+        domain_id = event.domain_id()
+        if domain_id is None:
+            raise AssertionError("Event domain ID is None")
+        await self._event_hub.close_by_alias(event.event_domain(), domain_id)

--- a/src/ai/backend/manager/services/container_registry/service.py
+++ b/src/ai/backend/manager/services/container_registry/service.py
@@ -48,7 +48,7 @@ class ContainerRegistryService:
 
             scanner_cls = get_container_registry_cls(registry_row)
             scanner = scanner_cls(self._db, registry_name, registry_row)
-            result = await scanner.rescan_single_registry(None)
+            result = await scanner.rescan_single_registry(action.progress_reporter)
 
         return RescanImagesActionResult(
             images=result.images, errors=result.errors, registry=registry_row.to_dataclass()

